src/inc/payment-gateways/ppcp-gateway.php

```diff
@@ -39,22 +39,99 @@ function get_from_order( $value, \WC_Order $order ) {
+ */
+function select_first_purchase_unit( $ppcp_data_order ) {
+	return select_first(
+		$ppcp_data_order,
+		fn( $obj ) => $obj?->purchase_units()
+	);
+}
+
+/**
+ * Select the first payments authorization or null if it could not be selected.
+ *
+ * @param WooCommerce\PayPalCommerce\ApiClient\Entity\Order $ppcp_data_order The PayPal Order object.
+ *
+ * @return WooCommerce\PayPalCommerce\ApiClient\Entity\Authorization|null The resulting Authorization if one is found.
+ */
+function select_first_payments_authorization( $ppcp_data_order ) {
+	return select_first(
+		select_first_purchase_unit( $ppcp_data_order ),
+		fn( $obj ) => $obj?->payments()?->authorizations()
+	);
+}
+
+/**
+ * Select the first payments caputure or null if it could not be selected.
+ *
+ * @param WooCommerce\PayPalCommerce\ApiClient\Entity\Order $ppcp_data_order The PayPal Order object.
+ *
+ * @return WooCommerce\PayPalCommerce\ApiClient\Entity\Capture|null The resulting Capture if one is found.
+ */
+function select_first_payments_capture( $ppcp_data_order ) {
+	return select_first(
+		select_first_purchase_unit( $ppcp_data_order ),
+		fn( $obj ) => $obj?->payments()?->captures()
+	);
+}
+
 add_filter( 'sift_for_woocommerce_ppcp-gateway_payment_method_details_from_order', __NAMESPACE__ . '\get_from_order', 10, 2 );
 add_filter( 'sift_for_woocommerce_ppcp-gateway_charge_details_from_order', __NAMESPACE__ . '\get_from_order', 10, 2 );
 
 add_filter( 'sift_for_woocommerce_ppcp-gateway_payment_type_string', fn( $value, $ppcp_gateway_payment_type ) => 'ppcp' === $ppcp_gateway_payment_type ? '$third_party_processor' : $value );
 
 add_filter( 'sift_for_woocommerce_ppcp-gateway_card_last4', fn( $value, $ppcp_data ) => $ppcp_data['wc_order']?->get_meta_data( PayPalGateway::FRAUD_RESULT_META_KEY )['card_last_digits'] ?? $value, 10, 2 );
-add_filter( 'sift_for_woocommerce_ppcp-gateway_avs_result_code', fn( $value, $ppcp_data ) => $ppcp_data['order']?->purchase_units()[0]?->payments()?->authorizations()[0]?->fraud_processor_response()->avs_code() ?? $value, 10, 2 );
-add_filter( 'sift_for_woocommerce_ppcp-gateway_cvv_result_code', fn( $value, $ppcp_data ) => $ppcp_data['order']?->purchase_units()[0]?->payments()?->authorizations()[0]?->fraud_processor_response()->cvv_code() ?? $value, 10, 2 );
-add_filter( 'sift_for_woocommerce_ppcp-gateway_verification_status', fn( $value, $ppcp_data ) => $ppcp_data['order']?->purchase_units()[0]?->payments()?->authorizations()[0]?->status()->name() ?? $value, 10, 2 );
-add_filter( 'sift_for_woocommerce_ppcp-gateway_decline_reason_code', fn( $value, $ppcp_data ) => $ppcp_data['order']?->purchase_units()[0]?->payments()?->authorizations()[0]?->to_array()['reason_code'] ?? $value, 10, 2 );
+add_filter( 'sift_for_woocommerce_ppcp-gateway_avs_result_code', fn( $value, $ppcp_data ) => select_first_payments_authorization( $ppcp_data['order'] )?->fraud_processor_response()->avs_code() ?? $value, 10, 2 );
```

Comment:
So

```
add_filter( 'sift_for_woocommerce_ppcp-gateway_avs_result_code', fn( $value, $ppcp_data ) => $ppcp_data['order']?->purchase_units()?[0]?->payments()?->authorizations()[0]?->fraud_processor_response()->avs_code() ?? $value
```

Would not work? I did not know...

Reply:
That was my first thought, but no - it's a syntax error. There is an alternative, but I'm not sure if we want to use it as it's a bit harder to read. (context: p1737057534401999-slack-C6JKZ8MLK):

```php
print_r( ( ( ( $ppcp_data['order'] ?? null )?->purchase_units() ?? [ null ] )[0]?->payments()?->authorizations() ?? [ null ] )[0]?->to_array()['reason_code'] ?? $value );
```
or
```php
print_r( ( ( ( $ppcp_data['order'] ?? null )?->purchase_units()[0] ?? null )?->payments()?->authorizations()[0] ?? null )?->to_array()['reason_code'] ?? $value );
```

(replacing `print_r` with the filter and function)